sockeye/encoder.py

```diff
@@ -413,3 +470,173 @@ def get_rnn_cells(self) -> List[mx.rnn.BaseRNNCell]:
+    :param num_filters: Number of filters of each width.
+    :param pool_stride: Stride for pooling layer after convolutions.
+    :param num_highway_layers: Number of highway layers for segment embeddings.
+    :param prefix: Name prefix for symbols of this encoder.
+    :param dropout: Dropout probability.
+    """
+
+    def __init__(self,
+                 num_embed: int,
+                 max_filter_width: int = 8,
+                 num_filters: List[int] = None,
+                 pool_stride: int = 5,
+                 num_highway_layers: int = 4,
+                 prefix: str = C.CHAR_SEQ_ENCODER_PREFIX,
+                 dropout: float = 0.):
+        if not num_filters:
+            num_filters = [200, 200, 250, 250, 300, 300, 300, 300]
+        assert len(num_filters) == max_filter_width, "num_filters must have max_filter_width elements."
+        self.num_embed = num_embed
+        self.max_filter_width = max_filter_width
+        self.num_filters = num_filters[:]
+        self.pool_stride = pool_stride
+        self.num_highway_layers = num_highway_layers
+        self.prefix = prefix
+        self.dropout = dropout
+        self.prefix = prefix
+
+        self.conv_weight = {filter_width: mx.sym.Variable("%s%s%d%s" % (self.prefix, "conv_", filter_width, "_weight"))
+                            for filter_width in range(1, self.max_filter_width + 1)}
+        self.conv_bias = {filter_width: mx.sym.Variable("%s%s%d%s" % (self.prefix, "conv_", filter_width, "_bias"))
+                          for filter_width in range(1, self.max_filter_width + 1)}
+
+        self.gate_weight = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "gate_", i, "_weight"))
+                            for i in range(self.num_highway_layers)]
+        self.gate_bias = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "gate_", i, "_bias"))
+                          for i in range(self.num_highway_layers)]
+
+        self.transform_weight = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "transform_", i, "_weight"))
+                                 for i in range(self.num_highway_layers)]
+        self.transform_bias = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "transform_", i, "_bias"))
+                               for i in range(self.num_highway_layers)]
+
+    def encode(self, data: mx.sym.Symbol, data_length: mx.sym.Symbol, seq_len: int) -> mx.sym.Symbol:
+        """
+        Encodes data given sequence lengths of individual examples and maximum sequence length.
+
+        :param data: Input data (batch_size, seq_len, num_embed).
+        :param data_length: Vector with sequence lengths.
+        :param seq_len: Maximum sequence length.
+        :return: Encoded input data.
+        """
+
+        total_num_filters = sum(self.num_filters)
+        encoded_seq_len = self.get_encoded_seq_len(seq_len)
+
+        # (batch_size, channel=1, seq_len, num_embed)
+        data = mx.sym.Reshape(data=data, shape=(-1, 1, seq_len, self.num_embed))
```

Comment:
MXNet's convolution and pooling layers are built for image processing (with much hard-coded behavior), so we make our sentences look like images.